src/raster/buffer.rs

```diff
@@ -0,0 +1,86 @@
+use crate::raster::GdalType;
+
+#[cfg(feature = "ndarray")]
+use ndarray::Array2;
+
+/// A 2-D array backed by it's `size` (cols, rows) and a row-major `Vec<T>` and it's dimensions.
+#[derive(Debug, Clone, PartialEq)]
+pub struct Buffer<T> {
+    pub size: (usize, usize),
+    pub data: Vec<T>,
+}
+
+impl<T: GdalType> Buffer<T> {
+    /// Construct a new buffer from `size` (`(cols, rows)`) and `Vec<T>`.
+    ///
+    /// # Notes
+    /// Assumes `size.0 * size.1 == data.len()`.
+    pub fn new(size: (usize, usize), data: Vec<T>) -> Buffer<T> {
+        debug_assert!(
+            size.0 * size.1 == data.len(),
+            "size {:?} does not match length {}",
+            size,
+            data.len()
+        );
+        Buffer { size, data }
+    }
+
+    #[cfg(feature = "ndarray")]
+    /// Convert `self` into an [`ndarray::Array2`].
+    pub fn to_array(self) -> crate::errors::Result<Array2<T>> {
+        // Array2 shape is (rows, cols) and Buffer shape is (cols in x-axis, rows in y-axis)
+        Ok(Array2::from_shape_vec(
+            (self.size.1, self.size.0),
+            self.data,
+        )?)
+    }
+}
+
+pub type ByteBuffer = Buffer<u8>;
+
+#[cfg(feature = "ndarray")]
+impl<T: GdalType> TryFrom<Buffer<T>> for Array2<T> {
+    type Error = crate::errors::GdalError;
+
+    fn try_from(value: Buffer<T>) -> Result<Self, Self::Error> {
+        value.to_array()
+    }
+}
+
+#[cfg(feature = "ndarray")]
+impl<T: GdalType + Copy> From<Array2<T>> for Buffer<T> {
+    fn from(value: Array2<T>) -> Self {
+        // Array2 shape is (rows, cols) and Buffer shape is (cols in x-axis, rows in y-axis)
+        let shape = value.shape().to_vec();
+        let data: Vec<T> = if value.is_standard_layout() {
+            value.into_raw_vec()
+        } else {
+            value.iter().copied().collect()
+        };
```

Comment:
I'm attempting here to efficiently convert back to a `Vec<T>` while protecting against receiving a column-major Array.

The old implementation just assumed the storage ordering was always row-major: 


https://github.com/georust/gdal/pull/494/files#diff-432e492d6a25cce04b77014813aea6861769ff42ce0c362d3f7850fb9e21ab5cL655

That said, I'm not sure how to test this properly.

Reply:
It might be easier to use [`as_standard_layout`](https://docs.rs/ndarray/latest/ndarray/struct.ArrayBase.html#method.as_standard_layout).

To test, I think we can use e.g. `Shape::from((2, 2)).set_f(true)` (from the `ShapeBuilder` trait).

Reply:
This is what it looks like now, which I would guess forces a copy regardless of storage ordering, but I'm not sure. At least in the case for `value.into_raw_vec()`, you're given ownership of the underlying buffer:

```rust
        let shape = value.shape();
        let (rows, cols) = (shape[0], shape[1]);
        let data = value
            .as_standard_layout()
            .iter()
            .copied()
            .collect::<Vec<T>>();
        Buffer::new((cols, rows), data)
```
https://github.com/georust/gdal/pull/494/files#diff-92db5ec20d7f8881c279099dfd10ceb3e36c1cec359f7d463d4f5a35497837f1R55-R62

Reply:
Yeah, I think the old version was better. Maybe use `value.into_raw_vec()` as before, and `.to_vec()` instead of `collect`?

Reply:
`to_vec` only exists on `Array1`.

https://docs.rs/ndarray/latest/ndarray/struct.ArrayBase.html#methods-for-1-d-arrays

Reply:
FWIW, still testing/working on this section; inspecting in debugger to follow what's happening.

Reply:
The tests:

https://github.com/georust/gdal/pull/494/files#diff-92db5ec20d7f8881c279099dfd10ceb3e36c1cec359f7d463d4f5a35497837f1R174-R191


Confirmed in debugger that with `set_f(true)` the execution path goes through the `! is_standard_layout` path.